Changes related to multithreading

lobis opened the pull request:
![lobis](https://badgen.net/badge/PR%20submitted%20by%3A/lobis/blue) ![Large: 3205](https://badgen.net/badge/PR%20Size/Large%3A%203205/red) [![](https://gitlab.cern.ch/rest-for-physics/restG4/badges/lobis-prepare-for-multithreading/pipeline.svg)](https://gitlab.cern.ch/rest-for-physics/restG4/-/commits/lobis-prepare-for-multithreading) [![](https://gitlab.cern.ch/rest-for-physics/geant4lib/badges/lobis-prepare-for-multithreading/pipeline.svg)](https://gitlab.cern.ch/rest-for-physics/geant4lib/-/commits/lobis-prepare-for-multithreading) [![](https://gitlab.cern.ch/rest-for-physics/framework/badges/lobis-prepare-for-multithreading/pipeline.svg)](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/lobis-prepare-for-multithreading) [![](https://github.com/rest-for-physics/restG4/actions/workflows/validation.yml/badge.svg?branch=lobis-prepare-for-multithreading)](https://github.com/rest-for-physics/restG4/commits/lobis-prepare-for-multithreading)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

* https://github.com/rest-for-physics/geant4lib/pull/59
* https://github.com/rest-for-physics/framework/pull/273

This PR introduces significant changes necessary to allow for multithreading to work. It also introduces some minor features that appeared naturally along the way.

README has also been updated with a description of some of these features.

`restG4` has new cli options: 
* `-t` to select the number of threads for multithreading
* `-n` to select the number of generator events (and override the value on the rml file) 
* `-e` to select the "desired number of entries" so that the simulation will stop once this number of entries is achieved. In practise probably a few more entries will be saved on the file. The value of `nEvents` will be updated to the corresponding value of the number of simulated events.
* `--time` to set a maximum simulation time. As the previous option the simulation will stop if more than the set time has elapsed. The value passed to this flag looks like: `1h30m20s`, `1m20s`, `1h`, `20s`, etc.

***Now the simulation can be stopped (via CTRL+C) and instead of crashing it will attempt to save results into disk, also updating the `nEvents` to the correct value.***

All examples have been updated to the changes described in https://github.com/rest-for-physics/geant4lib/pull/59 and a new example has been added to validate primary generators (anyone is welcome to write more checks in this example). As mentioned in the Geant4lib PR, RML files that use the old naming conventions (`energyDist` instead of `energy` etc.) should still work.

Now we are using Geant4 `SteppingVerbose` which allows to store the initial position of the track. This means a new hit type "Init" is saved for every track and means we can easily draw tracks. We don't need to find the last position of the previous track to draw the first point of the current track.

`StackingAction` is used to separate decays into subevents.

Logic has been simplified and methods such as `EventAction::ReOrderTrackIds` have been removed, which increases simulation speed.

After this PR multithreading should work (via the `-t` flag, such as `restG4 -t 4 simulation.rml`). Everything seems to work so far but the results have yet to be validated extensively. Using the same random seed with different number of threads will produce different results!

nkx111 commented:
I suggest to unify the options. In `restManager` we have `-j` to indicate threading, and `-e` to set desired number of entries.

There are also several options that I think would be very useful and can be added:
`-c`: set the rml file. We could use both `restG4 XXX.rml` and `restG4 -n 1000 -c XXX.rml -S 1000`
`-s`: set random seed. This would be useful in condor job uploading
`-v`: set verbose level

lobis commented:
> 



> I suggest to unify the options. In `restManager` we have `-j` to indicate threading, and `-e` to set desired number of entries.
> 
> There are also several options that I think would be very useful and can be added: `-c`: set the rml file. We could use both `restG4 XXX.rml` and `restG4 -n 1000 -c XXX.rml -S 1000` `-s`: set random seed. This would be useful in condor job uploading `-v`: set verbose level

I agree, I have modified options to match the ones in `restManager` and added option for seed.

juanangp commented on src/SimulationManager.cxx:
```diff
@@ -1,23 +1,553 @@
+        fPrimaryDirections.emplace_back(momentum.x(), momentum.y(), momentum.z());
+    }
+
+    return;
```
Check this `return`

juanangp commented:
I think we should re-visit the multithreading because I think that some variables should be defined as `atomic`. Also, I think the signal handler and the abort logic can be improved.
Anyhow, I think that this PR can be merged and we can do these changes later on, can you create an issue about that?
